<a href="https://colab.research.google.com/github/Daniel-Martinez24/cuenca-empleados/blob/main/cuenca.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preguntas


1.   ¿Cuantas personas trabajan en la cuenca?
2.  ¿Cuantas pertencen al ORGANISMO DE CUENCA LERMA SANTIAGO PACIFIC?

Nota: la posible respuesta de la segunda es 284 segun la revisión rapida



In [ ]:
!pip install ghostscript
!pip install camelot-py[cv]
!pip install excalibur-py
!apt install ghostscript python3-tk
from ctypes.util import find_library
print(find_library("gs")) #will display libgs.so.9 if installed; will print None if not
!excalibur initdb

# se importa camelot
import camelot

In [2]:
import pandas as pd

In [3]:
pdfCuencua = '/content/drive/MyDrive/proyectos/cuenca/Plantilla del personal.pdf'

In [4]:
tables = camelot.read_pdf(pdfCuencua, flavor='stream', pages='all')

In [5]:
# tiene 279 n pero el pdf tiene 274
tabla_final = tables[-1].df
tabla_final.drop(range(0,2),axis=0).reset_index(drop=True)

# tabla_final.rename(columns=tabla_final[1])

0           1  ...    4          5
0    ZEPEDA   HERNANDEZ  ...    7       BASE
1    ZEPEDA       LOPEZ  ...  M42  CONFIANZA
2    ZEPEDA     MENDOZA  ...  O23  CONFIANZA
3    ZEPEDA     SALAZAR  ...    6  CONFIANZA
4   ZERMEÑO      MENDEZ  ...    6       BASE
5    ZETINA      GARCIA  ...    7       BASE
6    ZETINA       PEREZ  ...    6       BASE
7    ZETINA     ROBLEDA  ...  N11  CONFIANZA
8    ZETINA      TORRES  ...    7       BASE
9    ZETINA      TORRES  ...    7       BASE
10   ZUMAYA   HERNANDEZ  ...    6       BASE
11   ZUMAYA        LEON  ...    6       BASE
12   ZUÑIGA      ALANIZ  ...    6       BASE
13   ZUÑIGA      ARENAS  ...    7       BASE
14   ZUÑIGA      ARVIZU  ...    6       BASE
15   ZUÑIGA   CASTAÑEDA  ...    6       BASE
16   ZUÑIGA    DE LEIJA  ...    6       BASE
17   ZUÑIGA       FRIAS  ...    7       BASE
18   ZUÑIGA       GOMEZ  ...    6       BASE
19   ZUÑIGA       GOMEZ  ...    6       BASE
20   ZUÑIGA    GONZALEZ  ...  O11  CONFIANZA
21   ZUÑIGA   HERNANDEZ  ...    6  CONFIANZA
22   ZUÑIGA     NAVARRO  ...    6       BASE
23   ZUÑIGA        NULL  ...    7       BASE
24   ZUÑIGA     PAREDES  ...    6       BASE
25   ZUÑIGA        PIÑA  ...    8       BASE
26   ZUÑIGA     RAMIREZ  ...    6       BASE
27   ZUÑIGA   RODRIGUEZ  ...    6  CONFIANZA
28   ZUÑIGA     SALAZAR  ...    6  CONFIANZA
29   ZUÑIGA     TARANGO  ...    7       BASE
30   ZUÑIGA     TARANGO  ...    7       BASE
31   ZUÑIGA      TELLEZ  ...    7       BASE
32   ZUÑIGA  VALDERRAMA  ...    7  CONFIANZA
33   ZUÑIGA       VELOZ  ...    7       BASE
34   ZURITA     ESTRADA  ...    6       BASE
35   ZURITA     ESTRADA  ...    7       BASE
36   ZURITA    MARTINEZ  ...    7       BASE
37   ZURITA     MENDOZA  ...    7       BASE
38   ZURITA       ROGEL  ...    8       BASE
39                LOPEZ  ...    7       BASE

[40 rows x 6 columns]

In [6]:
def limpiar_lista_dataframe(lista):
  '''
  Recibe la lista con DF y devuelve la misma lista con datos limpios
  '''
  lista_limpia = []
  for df in lista:
    lista_limpia.append(limpiar_df(df.df))
  return lista_limpia


In [7]:
def limpiar_df(df, espacios=range(0,2), cabecera=1):
  '''
  Devuelve el DF limpio
  '''
  nuevo_df = df.drop(espacios,axis=0).reset_index(drop=True)
  
  tabla_final.drop(range(0,2),axis=0).reset_index(drop=True)
  return nuevo_df

In [8]:
nueva_lista = limpiar_lista_dataframe(tables)
nueva_lista[-1]

0           1  ...    4          5
0    ZEPEDA   HERNANDEZ  ...    7       BASE
1    ZEPEDA       LOPEZ  ...  M42  CONFIANZA
2    ZEPEDA     MENDOZA  ...  O23  CONFIANZA
3    ZEPEDA     SALAZAR  ...    6  CONFIANZA
4   ZERMEÑO      MENDEZ  ...    6       BASE
5    ZETINA      GARCIA  ...    7       BASE
6    ZETINA       PEREZ  ...    6       BASE
7    ZETINA     ROBLEDA  ...  N11  CONFIANZA
8    ZETINA      TORRES  ...    7       BASE
9    ZETINA      TORRES  ...    7       BASE
10   ZUMAYA   HERNANDEZ  ...    6       BASE
11   ZUMAYA        LEON  ...    6       BASE
12   ZUÑIGA      ALANIZ  ...    6       BASE
13   ZUÑIGA      ARENAS  ...    7       BASE
14   ZUÑIGA      ARVIZU  ...    6       BASE
15   ZUÑIGA   CASTAÑEDA  ...    6       BASE
16   ZUÑIGA    DE LEIJA  ...    6       BASE
17   ZUÑIGA       FRIAS  ...    7       BASE
18   ZUÑIGA       GOMEZ  ...    6       BASE
19   ZUÑIGA       GOMEZ  ...    6       BASE
20   ZUÑIGA    GONZALEZ  ...  O11  CONFIANZA
21   ZUÑIGA   HERNANDEZ  ...    6  CONFIANZA
22   ZUÑIGA     NAVARRO  ...    6       BASE
23   ZUÑIGA        NULL  ...    7       BASE
24   ZUÑIGA     PAREDES  ...    6       BASE
25   ZUÑIGA        PIÑA  ...    8       BASE
26   ZUÑIGA     RAMIREZ  ...    6       BASE
27   ZUÑIGA   RODRIGUEZ  ...    6  CONFIANZA
28   ZUÑIGA     SALAZAR  ...    6  CONFIANZA
29   ZUÑIGA     TARANGO  ...    7       BASE
30   ZUÑIGA     TARANGO  ...    7       BASE
31   ZUÑIGA      TELLEZ  ...    7       BASE
32   ZUÑIGA  VALDERRAMA  ...    7  CONFIANZA
33   ZUÑIGA       VELOZ  ...    7       BASE
34   ZURITA     ESTRADA  ...    6       BASE
35   ZURITA     ESTRADA  ...    7       BASE
36   ZURITA    MARTINEZ  ...    7       BASE
37   ZURITA     MENDOZA  ...    7       BASE
38   ZURITA       ROGEL  ...    8       BASE
39                LOPEZ  ...    7       BASE

[40 rows x 6 columns]

In [9]:
'''
nueva_lista[125] == nueva_lista[124]

Al parecer esta libreria comete el error de añadir mas de una vez la misma información
Va segundo proceso de limpieza
'''
nueva_lista[125].equals(nueva_lista[124])

True

In [10]:
'''
Eliminar las partes repetidas
'''
position = 0
lista_sin_repetidos = []
for elemento in nueva_lista:
  if elemento.equals(nueva_lista[position + 1]):
      print(position)
  else:    
      lista_sin_repetidos.append(elemento)

  if position < 277:
    position += 1
  else:
    position = 0

100
107
124
178
210


In [16]:
lista_sin_repetidos[273].tail() # son los valores esperados :)

0         1  ...  4     5
35  ZURITA   ESTRADA  ...  7  BASE
36  ZURITA  MARTINEZ  ...  7  BASE
37  ZURITA   MENDOZA  ...  7  BASE
38  ZURITA     ROGEL  ...  8  BASE
39             LOPEZ  ...  7  BASE

[5 rows x 6 columns]

In [12]:
'''
por si las dudas voy a confirmar que ningun elemento se repite  XD
'''
position = 0
for elemento in lista_sin_repetidos:
  if elemento.equals(lista_sin_repetidos[position + 1]):
      pass
      print(position)
      print(position + 1)

  if position < len(lista_sin_repetidos)-2:
    position += 1

272
273


In [13]:
# concat
empleados_cuenca = pd.concat(lista_sin_repetidos).reset_index(drop=True)

Primera pregunta


In [17]:
empleados_cuenca

0         1                2  ...    4          5    6
0           .      MEZA  GREGORIA ESTELA  ...    7       BASE  NaN
1           .   VAZQUEZ           JOSEFA  ...    7       BASE  NaN
2        ABAD  CARRANZA        JOSE LUIS  ...  O23  CONFIANZA  NaN
3        ABAN  ESQUIVEL  ARMANDO EDUARDO  ...    7       BASE  NaN
4      ABARCA   RAMIREZ   ALFREDO ISMAEL  ...    8  CONFIANZA  NaN
...       ...       ...              ...  ...  ...        ...  ...
11228  ZURITA   ESTRADA     OSCAR ARTURO  ...    7       BASE  NaN
11229  ZURITA  MARTINEZ   MONICA CRISTEL  ...    7       BASE  NaN
11230  ZURITA   MENDOZA   VICTOR RICARDO  ...    7       BASE  NaN
11231  ZURITA     ROGEL     VICTOR ANGEL  ...    8       BASE  NaN
11232             LOPEZ    JOSE DE JESUS  ...    7       BASE  NaN

[11233 rows x 7 columns]

11233 rows se puede interpretar como 11233 personas

Segunda
ORGANISMO DE CUENCA LERMA SANTIAGO PACIFIC

In [42]:
lerma_santiago = empleados_cuenca.loc[empleados_cuenca[3].str.contains('812 ORGANISMO DE CUENCA LERMA SANTIAGO PACIFIC')]
lerma_santiago

0         1               2  ...    4          5    6
42      ACOSTA  GONZALEZ            LUIS  ...  M11  CONFIANZA  NaN
52      ACOSTA    OROZCO        FERNANDO  ...    7       BASE  NaN
53      ACOSTA    OROZCO       JOSE LUIS  ...    6       BASE  NaN
97      AGUAYO     ADAME       MARGARITA  ...    7  CONFIANZA  NaN
111    AGUILAR   BONILLA  GEMA GUADALUPE  ...    8       BASE  NaN
...        ...       ...             ...  ...  ...        ...  ...
11192   ZEPEDA     ANGEL        SALVADOR  ...    6       BASE  NaN
11195   ZEPEDA   MENDOZA  BLANCA MARIELA  ...  O23  CONFIANZA  NaN
11211   ZUÑIGA     GOMEZ          DANIEL  ...    6       BASE  NaN
11212   ZUÑIGA     GOMEZ           JULIA  ...    6       BASE  NaN
11219   ZUÑIGA   RAMIREZ  BLANCA LIZBETH  ...    6       BASE  NaN

[280 rows x 7 columns]

280 rows se puede interpretar como 280 valores